scripts/calculateRevenue/protocols/utils/networks.ts

```diff
@@ -0,0 +1,59 @@
+import { QueryResponse, TransactionField } from '@envio-dev/hypersync-client'
+import { getHyperSyncClient } from '../../../utils'
+import { NetworkId } from '../../../types'
+
+export async function fetchTotalGasUsed({
+  networkId,
+  users,
+  startBlock,
+  endBlock,
+}: {
+  networkId: NetworkId
+  users: string[]
+  startBlock?: number
+  endBlock?: number
+}): Promise<number> {
+  let fromBlock = startBlock ?? 0
+  let totalGasUsed = 0
+  let hasMoreBlocks = true
+
+  try {
+    const client = getHyperSyncClient(networkId)
+
+    const query = {
+      transactions: [{ from: users }],
+      fieldSelection: {
+        transaction: [TransactionField.GasUsed, TransactionField.GasPrice],
+      },
+      fromBlock,
+      ...(endBlock && { toBlock: endBlock }),
+    }
+
+    do {
+      const response: QueryResponse = await client.get(query)
+
+      if (
+        response.nextBlock <= fromBlock ||
+        !response.data.transactions.length
+      ) {
+        hasMoreBlocks = false
+      }
+
+      for (const tx of response.data.transactions) {
+        totalGasUsed += Number(tx.gasUsed ?? 0) * Number(tx.gasPrice ?? 0)
+      }
+
+      fromBlock = response.nextBlock
+      query.fromBlock = fromBlock
+
+      if (endBlock && fromBlock >= endBlock) {
```

Comment:
It seems like the intent of this check is to determine if we've reached "the end" of the results. In the case of `endBlock` not being set, this will never be true. Is `response.nextBlock` undefined when we've caught up with the current state of the chain? If so, can we change this to e.g. `(endBlock && fromBlock >= endBlock) || (!endBlock && !response.nextBlock))`?

Reply:
I fixed this in 0e858e96c6172a661ef39b7f3595cac62e8615b3 when `endBlock` is provided. If `endBlock` is missing, the function makes one additional request before exiting the loop. The missing `endBlock` shouldn’t happen since the CLI argument `endTimestamp` is required, but I've added a test for this regardless.